sdk/clientcore/core/src/main/java/io/clientcore/core/util/Union.java

```diff
@@ -0,0 +1,326 @@
+ * </pre>
+ * <!-- end io.clientcore.core.util.union.UnionJavaDocCodeSnippetsCollectionType -->
+ *
+ * <p><strong>Consume the value of the Union if it is of the expected type</strong></p>
+ *
+ * <!-- src_embed io.clientcore.core.util.union.UnionJavaDocCodeSnippetsSwitch -->
+ * <pre>
+ * Union union = Union.ofTypes&#40;String.class, Integer.class&#41;;
+ * union.setValue&#40;&quot;Hello&quot;&#41;;
+ * switch &#40;union.getValue&#40;&#41;&#41; &#123;
+ *     case String s -&gt; System.out.println&#40;&quot;String value: &quot; + s&#41;;
+ *     case Integer i -&gt; System.out.println&#40;&quot;Integer value: &quot; + i&#41;;
+ *     default -&gt; throw new IllegalArgumentException&#40;
+ *         &quot;Unknown type: &quot; + union.getCurrentType&#40;&#41;.getTypeName&#40;&#41;&#41;;
+ * &#125;
+ * </pre>
+ * <!-- end io.clientcore.core.util.union.UnionJavaDocCodeSnippetsSwitch -->
+ *
+ * or
+ *
+ * <!-- src_embed io.clientcore.core.util.union.UnionJavaDocCodeSnippetsLambda -->
+ * <pre>
+ * Union union = Union.ofTypes&#40;String.class, Integer.class&#41;;
+ * union.setValue&#40;&quot;Hello&quot;&#41;;
+ * union.tryConsume&#40;
+ *     v -&gt; System.out.println&#40;&quot;String value: &quot; + v&#41;, String.class&#41;;
+ * union.tryConsume&#40;
+ *     v -&gt; System.out.println&#40;&quot;Integer value: &quot; + v&#41;, Integer.class&#41;;
+ * </pre>
+ * <!-- end io.clientcore.core.util.union.UnionJavaDocCodeSnippetsLambda -->
+ *
+ */
+public final class Union {
+    private static final ClientLogger LOGGER = new ClientLogger(Union.class);
+
+    private final List<Type> types;
+    private Object value;
+    private Type currentType;
+
+    private Union(Type... types) {
+        if (types == null || types.length == 0) {
+            throw LOGGER.logThrowableAsError(new IllegalArgumentException("types cannot be null or empty"));
+        }
+
+        ArrayList<Type> typeCopy = new ArrayList<>(types.length);
+        for (int i = 0; i < types.length; i++) {
+            final Type currentType = types[i];
+            if (currentType == null) {
+                throw LOGGER.logThrowableAsError(
+                    new IllegalArgumentException("types cannot contain null values: null value in index " + i));
+            } else if (!(currentType instanceof Class<?> || currentType instanceof ParameterizedType)) {
+                throw LOGGER.logThrowableAsError(new IllegalArgumentException(
+                    String.format("types must be of type Class or ParameterizedType: type name is %s in index %d.",
+                        currentType.getTypeName(), i)));
+            }
+
+            typeCopy.add(types[i]);
+        }
+        this.types = typeCopy;
+    }
+
+    /**
+     * Creates a new instance of {@link Union} with the provided types.
+     *
+     * @param types The types of the union.
+     * @return A new instance of {@link Union}.
+     */
+    public static Union ofTypes(Type... types) {
```

Comment:
Given concerns around possible type erasure when using the APIs that get, set, and attempt to consume the value of the `Union`, would we want to limit the allowed `Type`s? 

Something along the lines of all `Class` instances are allowable and only `ParameterizedType`s are allowed. Where `ParameterizedType`s are further restricted to where the raw type can only be instances of `Collection` and `Map` as we should be able to handle those easily in a type erasure situation.

Reply:
Javadoc updated